Respond with TLS alert "bad_certificate" if certificate parsing fails

markus-jung-vivavis opened the pull request:
Improve the error handling in case a invalid (syntactically incorrect) certificate is received in a TLS handshake from the server. 

Previously, the exception thrown during ASN.1 parsing was not caught and therefore caused a TLS alert "internal_error". This PR handles the exception to respond with a RFC-compliant TLS alert "bad_certificate".

ottoka committed:
ProvTlsServer should use custom DH groups when configured

(cherry picked from commit 2a5b8858a0789c33d70de237ef0379db30ffd6e7)

yuhh0328 committed:
Updates : add tls-kyber

dghgit committed:
minor cleanup

dghgit committed:
removed external dependency on provider composite name class
standardised names to RFC draft.

dghgit committed:
256 bit NTRU parameters

dghgit committed:
added support for NTRU HRSS 1373

dghgit committed:
minor refactoring - removal of common code.

dghgit committed:
added additional NTRU OIDs to provider.

dghgit committed:
change to use this.read() relates to github #1577

dghgit committed:
Merge branch 'bcgit:main' into main

dghgit committed:
fixed license reference, minor adjustments

dghgit committed:
minor wording, added mention of FIPS

dghgit committed:
Merge branch 'memoable' of https://github.com/seungyeonpark/bc-java into seungyeonpark-memoable

dghgit committed:
Merge branch 'seungyeonpark-memoable'

peterdettman committed:
Further checks in bcpg.sig

peterdettman committed:
EdDSA: Explicit guard against infinite looping

peterdettman committed:
Remove duplicate test case

dghgit committed:
fixed dodgy test...

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
Initial cut for id-alg-cek-hkdf-sha256 CMS support.

dghgit committed:
added import of java.logging (github #1608)

dghgit committed:
added new RSA mode for better TLS unwrap operation relates to github #1528

dghgit committed:
widened RSA private key support - relates to github #1528

dghgit committed:
corrected repackaging issue.

dghgit committed:
added missing copy statement

dghgit committed:
removed AllTests for keybox test

mwcw committed:
Added running ant build to CI.

mwcw committed:
Added ant to path.

mwcw committed:
Update to pass multiple args to javadoc

Commit:
Added call to gitlab_wait

Commit:
Removed gitlab_wait.sh from jobs.

dghgit committed:
added some explanatory comments.

dghgit committed:
minor changes to messages

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

peterdettman committed:
Round out implementation of TLS RSA PreMasterSecret

peterdettman committed:
Refactoring

dghgit committed:
removed use of PQC provider.

Commit:
NTRU KEMspi for jdk 21

Commit:
Merge branch 'tls-kyber' of https://github.com/samsungsds-research-papers/bc-java into samsungsds-research-papers-tls-kyber

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
digest name fix.
added jdk1.5 compatibility classes.

dghgit committed:
fixed checkstyle issue

dghgit committed:
reverted PGPUtil inclusion

peterdettman committed:
Fix various bitstring usages

peterdettman committed:
Fix engineGetOutputSize and add some TODOs

cipherboy committed:
Add regression tests for Ed448 loop

See also: 02f1fb441b6b72826ea892aea4a741403d9562f3

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

peterdettman committed:
Update EdDSA tests

dghgit committed:
corrected test to avoid failure where last byte was already zero.

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

Commit:
Added Mock Tls Tests

Commit:
Renamed kyber to mlkem

Commit:
Added mismatch test for different version of mlkem

Commit:
Added missing input validation for ML-KEM Encapsulation (fips 203 6.2)

Commit:
Replaced RuntimeException with IllegalArgumentException

dghgit committed:
minor changes for compatibility

dghgit committed:
minor changes for compatibility

dghgit committed:
added AES CBC based PKCS12 store and AES GCM based PKCS12 store

dghgit committed:
minor changes for compatibility

dghgit committed:
added use of Strings.toLowerCase()

peterdettman committed:
BCJSSE: Private key may not support encoding

peterdettman committed:
BCJSSE: Improved workaround for InetAddress limitation

- URLConnectionUtil now calls BCSSLSocket.setHost instead of direct SNI config

Commit:
Added KEM check in selectKeyShare/Group
minor name change

Commit:
added fail in catch

Commit:
quick little arrangement

peterdettman committed:
TLS: Rework ML-KEM PR

Commit:
Merge remote-tracking branch 'origin/samsungsds-research-papers-tls-kyber'

# Conflicts:
#	tls/src/main/java/org/bouncycastle/jsse/provider/NamedGroupInfo.java
#	tls/src/main/java/org/bouncycastle/tls/NamedGroup.java
#	tls/src/main/java/org/bouncycastle/tls/TlsServerProtocol.java
#	tls/src/main/java/org/bouncycastle/tls/TlsUtils.java
#	tls/src/main/java/org/bouncycastle/tls/crypto/TlsCrypto.java
#	tls/src/main/java/org/bouncycastle/tls/crypto/impl/bc/BcTlsCrypto.java
#	tls/src/main/java/org/bouncycastle/tls/crypto/impl/jcajce/JcaTlsCrypto.java

Commit:
Merged tests to tls ml-kem

Commit:
removed extra files due to naming

Commit:
removed extra files due to naming

dghgit committed:
added role back to simpler match for certificate matching (issue around locating trust anchors).

dghgit committed:
minor formatting

dghgit committed:
minor formatting

dghgit committed:
moved to 1.78, BCJSSE 1.0.19.

dghgit committed:
minor reformatting

dghgit committed:
compatibility updates.

mwcw committed:
Ensure test data changes are pulled into container.

mwcw committed:
Removed test data pull.

peterdettman committed:
Cleanup deprecated TLS code

peterdettman committed:
DTLS: more robust exclusion of stream ciphers

peterdettman committed:
TLS: Use more appropriate method to check for ECDH curve

dghgit committed:
Java 5 compatibility updates

dghgit committed:
Java 5 compatibility updates

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
Java 5 compatibility updates.

dghgit committed:
Java 5 compatibility updates.

dghgit committed:
Java 5 compatibility updates.

cipherboy committed:
Add release notes for 1.78

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

cipherboy committed:
Fix CCM input length check

As reported by @CipherGato, CCM fails to decrypt a full-length message
for n=13, q=2 (maximum length 65,535). This is because the input message
already includes the flags and nonce components (B0 from SP800 38C).
Correctly handle full-length CCM decryption.

Resolves: #1570

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

dghgit committed:
Java 5 compatibility

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
Java 4 compatibility updates.
Removed Camellia wrap enable property from OpenPGP (not required).

mtgag committed:
added name resolving for jurisdiction{C,ST,L}. These are used in EV certificates.

cipherboy committed:
Add CCM fix to release notes

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

dghgit committed:
Merge branch 'mtgag-jurisdiction_strings'

dghgit committed:
Java 4 updates

peterdettman committed:
Update release notes

dghgit committed:
Java 4 updates

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

dghgit committed:
minor cleanups

dghgit committed:
moved to 1.78

dghgit committed:
corrected version number

dghgit committed:
almost final releasenotes

dghgit committed:
update

dghgit committed:
Java 2, Java 3, compatibility updates

dghgit committed:
Java 1.1 Updates

dghgit committed:
rolled back accidental commit.

dghgit committed:
rolled back accidental commit.

dghgit committed:
rolled back accidental commit.

peterdettman committed:
Add SetHostSocketFactory constructor directly from host string

peterdettman committed:
Add PGPPublicKey.hasFingerprint method (reduce allocations)

peterdettman committed:
Refactor to use Pack

peterdettman committed:
TLS: Some work on GOST support (RFC 9189)

dghgit committed:
made class more forgiving on ASN.1 encodings

dghgit committed:
OSGi patches - added versioning, verified private packages

dghgit committed:
marked as 1.78.1

dghgit committed:
added missing requires

dghgit committed:
fixed JCA digest creation fallback

dghgit committed:
removed use of parameters class

dghgit committed:
fixed misspelt method, refactored out some dependent code into correct subpackages

dghgit committed:
removed import of internal classes

dghgit committed:
added empty extension block test (OCSP)

peterdettman committed:
Add exception detail message

dghgit committed:
added 1.78.1 details

dghgit committed:
tightened up version compatibility for OSGi

dghgit committed:
added maxVersion restriction number

dghgit committed:
moved to 1.78.1

Commit:
Merge branch '1.78.1' into 'main'

1.78.1

See merge request root/bc-java!11

dghgit committed:
added fallback tests for digest creation in PGP JCA layer

dghgit committed:
Java 5 to Java 8 BND script

dghgit committed:
started move to 1.79

KarolinHem committed:
Merge branch 'bcgit:main' into main

peterdettman committed:
Refactoring in tls

dghgit committed:
added jakarta dependency to OSGi header

dghgit committed:
Merge branch 'main' of gitlab.cryptoworkshop.com:root/bc-java

vanitasvitae committed:
Add utility methods to StreamUtil

vanitasvitae committed:
Add HexDumpUtil for formatted test output

vanitasvitae committed:
Add AbstractPacketTest containing DSL for packet-related tests

vanitasvitae committed:
Implement v6 packet parsing for OPS and SignaturePackets

vanitasvitae committed:
Avoid MPI encoding for Ed25519,Ed448 signatures

vanitasvitae committed:
Add tests for v6 OPS and SignaturePackets

dghgit committed:
initial LICENSE.md

KarolinHem committed:
Update CONTRIBUTING.md

KarolinHem committed:
Delete .github/ISSUE_TEMPLATE directory

KarolinHem committed:
Delete pull_request_template.md

KarolinHem committed:
Changed formatting in security problem Notes

KarolinHem committed:
Updated Notes on Security Problems

peterdettman committed:
Improved instanceof check

peterdettman committed:
EC updates from bc-csharp

dghgit committed:
added missing item about PEM parsing change 1.78

mwcw committed:
Added OSGI version creation when building snapshots.

dghgit committed:
Merge branch 'main' of github.com:KarolinHem/bc-java into KarolinHem-main

dghgit committed:
Merge branch 'KarolinHem-main'

dghgit committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

dghgit committed:
changed github relative links to explicit

peterdettman committed:
Refactoring (reduce allocations)

vanitasvitae committed:
Expose version, fingerprint and salt in PGPOnePassSignature

vanitasvitae committed:
Remove unused import

vanitasvitae committed:
Fix checkstyle issues

vanitasvitae committed:
Add FingerprintUtil helper for deriving key-ids from fingerprints

vanitasvitae committed:
Add HashUtils helper for looking up salt sizes for digest algorithms

vanitasvitae committed:
Fix formatting of tests

dghgit committed:
added CRLsign check

dghgit committed:
added CRLsign check

dghgit committed:
Merge branch 'sigv6_2' of https://github.com/pgpainless/bc-java into pgpainless-sigv6_2

dghgit committed:
Merge branch 'pgpainless-sigv6_2'

dghgit committed:
minor reorg of test classes - relates to github #1634

dghgit committed:
Merge remote-tracking branch 'refs/remotes/origin/main'

dghgit committed:
added default setting for CRL sign check

dghgit committed:
added default providing isOverride()

dghgit committed:
added use of default on CRL sign check.

dghgit committed:
cleared up some colliding tests

dghgit committed:
Merge branch 'fingerprintUtil' of https://github.com/pgpainless/bc-java into pgpainless-fingerprintUtil

dghgit committed:
Merge branch 'pgpainless-fingerprintUtil'

dghgit committed:
fixed checkstyle issues.

dghgit committed:
added FingerprintUtilTest (relates to github #1636)

dghgit committed:
bought util build in line with gradle util build

dghgit committed:
Merge branch 'hashUtil' of https://github.com/pgpainless/bc-java into pgpainless-hashUtil

dghgit committed:
Merge branch 'pgpainless-hashUtil'

vanitasvitae committed:
Use FingerprintUtil for deriving key-ID from fingerprint

vanitasvitae committed:
Use FingerprintUtil in test

dghgit committed:
added CVE ID for unallocated CVE

mwcw committed:
Added OSGI scan check to "check" stage

Commit:
swapped sha2_128f and sha2_128s for SPHINCSPlusParameterSpec.java (#1648)

vanitasvitae committed:
OpenPGP Packets: Remember packetTagFormat to allow round-tripping packets unmodified

vanitasvitae committed:
OpenPGP Packets: Remember packetTagFormat to allow round-tripping packets unmodified

vanitasvitae committed:
Refer to public key algorithm registries

vanitasvitae committed:
Add documentation to individual PK algorithm tags

vanitasvitae committed:
Deprecated RSA_SIGN, RSA_ENCRYPT, ELGAMAL_GENERAL

vanitasvitae committed:
Add new PK algorithm tags AEDH, AEDSA

vanitasvitae committed:
Document SymmetricKeyAlgorithmTags

vanitasvitae committed:
Document HashAlgorithmTags

vanitasvitae committed:
Document CompressionAlgorithmTags

vanitasvitae committed:
Add documentation for AEADAlgorithmTags

peterdettman committed:
Fix javadoc

peterdettman committed:
Refactor PEMParser

vanitasvitae committed:
Add documentation to ECDH,ECDSA,EdDSA BCPGKeys

vanitasvitae committed:
Add test for EC curve key conversion

peterdettman committed:
Fix KEMRecipientInfo sequence size restriction

peterdettman committed:
Improve KeyAgreeRecipientIdentifier getInstance methods

vanitasvitae committed:
Use Pack class for writing and reading key-IDs from byte arrays

vanitasvitae committed:
Fix conversion of ECDSA keys

vanitasvitae committed:
Add tests for conversion between JCA/BC keys

vanitasvitae committed:
Fix PGPKeyConverter classes support for X448,Ed448,X25519,Ed25519

vanitasvitae committed:
Document reversed MPI encoding / little-endian native encoding

vanitasvitae committed:
Fix javadoc reference to ECSecretBCPGKey

vanitasvitae committed:
Add error messages to Curve25519PrivateKeyEncodingTest

dghgit committed:
Merge branch 'documentSKTags' of https://github.com/pgpainless/bc-java into pgpainless-documentSKTags

dghgit committed:
Merge branch 'pgpainless-documentSKTags'

dghgit committed:
Merge branch 'documentPKTags' of https://github.com/pgpainless/bc-java into pgpainless-documentPKTags

dghgit committed:
Merge branch 'pgpainless-documentPKTags'

dghgit committed:
Merge branch 'documentHashTags' of https://github.com/pgpainless/bc-java into pgpainless-documentHashTags

dghgit committed:
Merge branch 'pgpainless-documentHashTags'

dghgit committed:
Merge branch 'documentCompressionTags' of https://github.com/pgpainless/bc-java into pgpainless-documentCompressionTags

dghgit committed:
Merge branch 'pgpainless-documentCompressionTags'

dghgit committed:
Merge branch 'documentAEADTags' of https://github.com/pgpainless/bc-java into pgpainless-documentAEADTags

dghgit committed:
Merge branch 'pgpainless-documentAEADTags'

dghgit committed:
Merge branch 'documentBCPGKeys' of https://github.com/pgpainless/bc-java into pgpainless-documentBCPGKeys

dghgit committed:
Merge branch 'pgpainless-documentBCPGKeys'

dghgit committed:
Merge branch 'roundtripPacketFormat' of https://github.com/pgpainless/bc-java into pgpainless-roundtripPacketFormat

dghgit committed:
Merge branch 'pgpainless-roundtripPacketFormat'

dghgit committed:
removed unnecessary import

vanitasvitae committed:
Fix OnePassSignaturePacketTest by only comparing encoding if format matches

dghgit committed:
Merge branch 'roundtripPacketFormat' of https://github.com/pgpainless/bc-java into pgpainless-roundtripPacketFormat

dghgit committed:
Merge branch 'pgpainless-roundtripPacketFormat'

dghgit committed:
Merge branch 'useFingerprintUtil' of https://github.com/pgpainless/bc-java into pgpainless-useFingerprintUtil

dghgit committed:
Merge branch 'pgpainless-useFingerprintUtil'

dghgit committed:
Merge branch 'fixKeyConversion' of https://github.com/pgpainless/bc-java into pgpainless-fixKeyConversion

dghgit committed:
Merge branch 'pgpainless-fixKeyConversion'

vanitasvitae committed:
Rename named curve lookup method

dghgit committed:
Merge branch 'pgpainless-fixECKeyConversion'

vanitasvitae committed:
Do not register BouncyCastleProvider in ECDSAKeyPairTest

vanitasvitae committed:
Fix legacy Ed448 signature field parsing

vanitasvitae committed:
Fix signing using legacy Ed448 keys

vanitasvitae committed:
Add signing/verification tests for LegacyEd25519/LegacyEd448 keys

cipherboy committed:
Improve Dilithium, LMS, SPHINCS+'s constant time behavior

Signed-off-by: Alexander Scheel <alexander.scheel@keyfactor.com>

dghgit committed:
added UTF8 support

dghgit committed:
Merge branch 'fixECKeyConversion' of https://github.com/pgpainless/bc-java

dghgit committed:
Generalized EC key processing for different providers, added use of X9.62 parameters for curve OID finding (relates to github #1671)

dghgit committed:
Minor cleanup of unused argument.

peterdettman committed:
Followup changes for custom DH groups

peterdettman committed:
TLS: Improve default server DH group selection

vanitasvitae committed:
Prevent conversion issue for EdDSA keys with leading 0

vanitasvitae committed:
Test conversion of Ed25519, Ed448 keys with leading zero

dghgit committed:
Merge branch 'ed_leading_zero' of https://github.com/pgpainless/bc-java into pgpainless-ed_leading_zero

dghgit committed:
Merge branch 'pgpainless-ed_leading_zero'

dghgit committed:
Merge branch 'fixEd448Signing' of https://github.com/pgpainless/bc-java into pgpainless-fixEd448Signing

dghgit committed:
Merge branch 'pgpainless-fixEd448Signing'

dghgit committed:
minor refactor

markus-jung-vivavis committed:
Respond with TLS alert "bad_certificate" if certificate parsing fails

peterdettman commented:
The basic idea is good, but the patch is a bit simplistic:

- we should deal also with processing of the client certificate
- shouldn't swallow existing TlsFatalAlert (which is an IOException subclass)
- probably not all errors from Certificate.parse should be bad_certificate; I guess decode_error should be used for some cases and there may be others.

markus-jung-vivavis committed:
Refine response to broken certificates received during TLS handshake

- Refactor response into common static method
- Apply response to certificates received by server, too
- Only throw a new TlsFatalAlert in case of an unspecific IOException
